File: libs/rust/manifests/src/checkpoint.rs
// Copyright © 2026 Mindclade, LLC. All Rights Reserved.
// Mindclade Proprietary and Confidential.
// SPDX-License-Identifier: LicenseRef-Mindclade-Proprietary
//

use crate::ArtifactRef;
use mindclade_content_digest::Digest;
use mindclade_faults::{Fault, FaultResult};
use std::collections::BTreeMap;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CheckpointComponentRef {
    pub name: String,
    pub artifact: ArtifactRef,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct DistributedCheckpointManifest {
    pub schema_version: u32,
    pub step: u64,
    pub world_size: u32,
    pub topology_digest: Digest,
    pub components: BTreeMap<String, ArtifactRef>,
}

impl DistributedCheckpointManifest {
    pub fn validate(&self) -> FaultResult<()> {
        if self.schema_version == 0 || self.world_size == 0 || self.components.is_empty() {
            return Err(Fault::invalid_argument(
                "distributed checkpoint manifest is incomplete",
            ));
        }
        for (name, a) in &self.components {
            if name.is_empty() || name.len() > 256 {
                return Err(Fault::invalid_argument(
                    "checkpoint component name is invalid",
                ));
            }
            a.validate()?;
        }
        Ok(())
    }
}

File: data/loaders/streaming/checkpoint.py
# Copyright © 2026 Mindclade, LLC. All Rights Reserved.
# Mindclade Proprietary and Confidential.
# SPDX-License-Identifier: LicenseRef-Mindclade-Proprietary
#

"""Immutable manifest-bound streaming position."""

from __future__ import annotations

import re
from dataclasses import dataclass

_DIGEST = re.compile(r"sha256:[0-9a-f]{64}")


@dataclass(frozen=True, slots=True)
class StreamCheckpoint:
    manifest_digest: str
    next_index: int
    epoch: int
    shuffle_seed: int

    def __post_init__(self) -> None:
        if not _DIGEST.fullmatch(self.manifest_digest):
            raise ValueError("stream checkpoint manifest digest is invalid")
        for value, name in (
            (self.next_index, "next_index"),
            (self.epoch, "epoch"),
            (self.shuffle_seed, "shuffle_seed"),
        ):
            if isinstance(value, bool) or not isinstance(value, int) or value < 0:
                raise ValueError(f"stream checkpoint {name} is invalid")
